Use scripts to initialize LogixNG variables

danielb987 opened the pull request:
This PR adds the possibility to use scripts to initialize local and global variables in LogixNG.

Scripts have three main advantages:
* They let you create an instance of an arbitrary Java object, for example if you need a variable which is a HashSet.
* They let you assign a named bean to a variable, for example sensors.provide("MySensor")
* Using a script file, you can do a more complex initialization, for example assign a map with a predefined set of values.

The PR also changes the initialization of global variables. It now first loads all the logixngs, modules and tables and then initializes the global variables. It allows a global variable to access a table during initialization.

danielb987 commented:
The attached example panel file demonstrates this PR.
[DS.LogixNG_GlobalVariables.txt](https://github.com/JMRI/JMRI/files/8585775/DS.LogixNG_GlobalVariables.txt)

The global variable `sensorBlink` is set to the sensor `BlinkSensor`. The variable is then used to turn the sensor active/inactive.

The global variable `exampleTableVariable` is set to a map with the `Turnout` column of the LogixNG table `MyTable`.

The global variable `exampleTableVariable2` is set to a map of maps with all the cells of the LogixNG table `MyTable`.

See the log for the printout of the different global variables.

danielb987 commented:
All scripts run by LogixNG, for example the action Script, gives the script access to these managers:
```
logixngs -> LogixNG_Manager
conditionalngs -> ConditionalNG_Manager
globalVariables -> GlobalVariableManager
logixngModules -> ModuleManager
logixngTables -> NamedTableManager
analogActions -> AnalogActionManager
analogExpressions -> AnalogExpressionManager
digitalActions -> DigitalActionManager
digitalBooleanActions -> DigitalBooleanActionManager
digitalExpressions -> DigitalExpressionManager
stringActions -> StringActionManager
stringExpressions -> StringExpressionManager
```
This works similar to the variables `sensors`, `turnouts`, and so on. Example:
`myTable = logixngTables.getNamedTable(tableName)`

Also, scripts run by the action or expression Script, also have access to the variable `symbolTable` that gives access to the current local variables. See: `jmri.jmrit.logixng.SymbolTable`.

danielb987 commented:
When a script expression is used to initialize the global variable, only the expression itself should be given.
Example: `sensors.provide("BlinkSensor")`

When a script file is used, the script works like any Jython script. The script tells which value to assign to the global variable by assigning the value to the variable `variable`.
Example: `variable.set(myMap)`

dsand47 commented:
Typing script content into the local variable window is difficult due to the rigid nature of the window.

![Screen Shot 2022-04-28 at 3 26 28 PM](https://user-images.githubusercontent.com/17280992/165840008-c45c625d-6a97-4330-b462-2184dcf24816.png)

The table content should expand when the window is made wider.

danielb987 commented:
> The sample fails.

I tried to download the panel file I uploaded before in this PR and it works fine for me. Can you try again and ensure that your branch from this PR is up-to-date? Thanks.

danielb987 commented:
> Typing script content into the local variable window is difficult due to the rigid nature of the window.

It turned out that I had a scrollpane in a JPanel in the contentpane. By removing the JPanel, the scrollpane now expands as the dialog is resized. So it was easy to fix.

dsand47 commented:
The failure is caused by a timing/sequencing issue.

```
     [java] 18:29:32,562 configurexml.LoadXmlConfigAction      INFO  - Loading selected file: /Users/das/JMRI/_Profiles/LNG.jmri/DS Test.xml [AWT-EventQueue-0]
     [java] 18:29:33,817 actions.LogData                       WARN  - ---- Start :: IQ:AUTO:0001 :: IQC:AUTO:0002 [JMRI LogixNGThread]
     [java] 18:29:33,818 actions.LogLocalVariables             WARN  - Log local variables: [JMRI LogixNGThread]
     [java] 18:29:33,819 actions.LogLocalVariables             WARN  -     Name: time, Value: null [JMRI LogixNGThread]
     [java] 18:29:33,819 actions.LogLocalVariables             WARN  - Global variables: [JMRI LogixNGThread]
     [java] 18:29:33,820 actions.LogLocalVariables             WARN  -     Global Name: gHour, value: 12 [JMRI LogixNGThread]
     [java] 18:29:33,820 actions.LogLocalVariables             WARN  -     Global Name: gMin, value: 0 [JMRI LogixNGThread]
     [java] 18:29:33,820 actions.LogLocalVariables             WARN  -     Global Name: timeMap, value: {} [JMRI LogixNGThread]
     [java] 18:29:33,820 actions.LogLocalVariables             WARN  -     Global Name: timeArray, value: [, , ] [JMRI LogixNGThread]
     [java] 18:29:33,821 actions.LogLocalVariables             WARN  -     Global Name: mySensor, value: null [JMRI LogixNGThread]
     [java] 18:29:33,821 actions.LogLocalVariables             WARN  -     Global Name: sensorBlink, value: null [JMRI LogixNGThread]
     [java] 18:29:33,821 actions.LogLocalVariables             WARN  -     Global Name: exampleTableVariable, value: null [JMRI LogixNGThread]
     [java] 18:29:33,821 actions.LogLocalVariables             WARN  -     Global Name: exampleTableVariable2, value: null [JMRI LogixNGThread]
     [java] 18:29:33,821 actions.LogLocalVariables             WARN  - Log local variables done [JMRI LogixNGThread]
     [java] 18:29:33,856 actions.DigitalFormula                INFO  - execute: timeArray[2] = format("The array time is %02d:%02d", timeArray[0], timeArray[1]) [JMRI LogixNGThread]
     [java] 0002 Table: IQT:AUTO:0001
     [java] 0003 Table: IQT:AUTO:0001
     [java] 18:29:38,949 implementation.DefaultConditionalNG   WARN  - ConditionalNG IQC:AUTO:0002 was aborted during execute: java.util.IllegalFormatConversionException: d != java.lang.String [JMRI LogixNGThread]
     [java] java.util.IllegalFormatConversionException: d != java.lang.String
     [java] 	at java.base/java.util.Formatter$FormatSpecifier.failConversion(Formatter.java:4426)
     [java] 	at java.base/java.util.Formatter$FormatSpecifier.printInteger(Formatter.java:2938)
     [java] 	at java.base/java.util.Formatter$FormatSpecifier.print(Formatter.java:2892)
     [java] 	at java.base/java.util.Formatter.format(Formatter.java:2673)
     [java] 	at java.base/java.util.Formatter.format(Formatter.java:2609)
     [java] 	at java.base/java.lang.String.format(String.java:2897)
     [java] 	at jmri.jmrit.logixng.util.parser.functions.StringFunctions$FormatFunction.calculate(StringFunctions.java:80)
     [java] 	at jmri.jmrit.logixng.util.parser.ExpressionNodeFunction.calculate(ExpressionNodeFunction.java:35)
     [java] 	at jmri.jmrit.logixng.util.parser.ExpressionNodeAssignmentOperator.calculate(ExpressionNodeAssignmentOperator.java:234)
     [java] 	at jmri.jmrit.logixng.actions.DigitalFormula.execute(DigitalFormula.java:147)
     [java] 	at jmri.jmrit.logixng.implementation.DefaultMaleDigitalActionSocket.execute(DefaultMaleDigitalActionSocket.java:45)
     [java] 	at jmri.jmrit.logixng.tools.debugger.DebuggerMaleDigitalActionSocket.execute(DebuggerMaleDigitalActionSocket.java:35)
     [java] 	at jmri.jmrit.logixng.implementation.DefaultFemaleDigitalActionSocket.execute(DefaultFemaleDigitalActionSocket.java:37)
     [java] 	at jmri.jmrit.logixng.actions.DigitalMany.execute(DigitalMany.java:116)
     [java] 	at jmri.jmrit.logixng.implementation.DefaultMaleDigitalActionSocket.execute(DefaultMaleDigitalActionSocket.java:45)
     [java] 	at jmri.jmrit.logixng.tools.debugger.DebuggerMaleDigitalActionSocket.execute(DebuggerMaleDigitalActionSocket.java:35)
     [java] 	at jmri.jmrit.logixng.implementation.DefaultFemaleDigitalActionSocket.execute(DefaultFemaleDigitalActionSocket.java:37)
     [java] 	at jmri.jmrit.logixng.implementation.DefaultConditionalNG.internalExecute(DefaultConditionalNG.java:153)
     [java] 	at jmri.jmrit.logixng.implementation.DefaultConditionalNG$ExecuteTask.run(DefaultConditionalNG.java:186)
     [java] 	at jmri.jmrit.logixng.util.LogixNG_Thread.lambda$new$0(LogixNG_Thread.java:143)
     [java] 	at java.base/java.lang.Thread.run(Thread.java:829)
     [java] 18:29:38,953 actions.LogData                       WARN  - ---- Start :: IQ:AUTO:0002 :: IGC:AUTO:0004 [JMRI LogixNGThread]
     [java] 18:29:39,006 actions.DigitalFormula                INFO  - execute: timeArray[0] = 0 [JMRI LogixNGThread]
     [java] 18:29:39,007 actions.DigitalFormula                INFO  - execute: timeArray[1] = 0 [JMRI LogixNGThread]
     [java] 18:29:39,007 actions.DigitalFormula                INFO  - execute: timeMap{"theHour"} = 0 [JMRI LogixNGThread]
     [java] 18:29:39,007 actions.DigitalFormula                INFO  - execute: timeMap{"theMin"} = 0 [JMRI LogixNGThread]
     [java] 18:29:39,007 actions.LogLocalVariables             WARN  - Log local variables: [JMRI LogixNGThread]
     [java] 18:29:39,007 actions.LogLocalVariables             WARN  - Global variables: [JMRI LogixNGThread]
     [java] 18:29:39,008 actions.LogLocalVariables             WARN  -     Global Name: gHour, value: 12 [JMRI LogixNGThread]
     [java] 18:29:39,008 actions.LogLocalVariables             WARN  -     Global Name: gMin, value: 0 [JMRI LogixNGThread]
     [java] 18:29:39,008 actions.LogLocalVariables             WARN  -     Global Name: timeMap, value: {theHour=0, theMin=0} [JMRI LogixNGThread]
     [java] 18:29:39,009 actions.LogLocalVariables             WARN  -     Global Name: timeArray, value: [0, 0, ] [JMRI LogixNGThread]
     [java] 18:29:39,009 actions.LogLocalVariables             WARN  -     Global Name: mySensor, value: IS1 [JMRI LogixNGThread]
     [java] 18:29:39,010 actions.LogLocalVariables             WARN  -     Global Name: sensorBlink, value: IS2 [JMRI LogixNGThread]
     [java] 18:29:39,010 actions.LogLocalVariables             WARN  -     Global Name: exampleTableVariable, value: {First siding=IT15, Left=Left turnout, Second siding=IT22, Right=Right turnout} [JMRI LogixNGThread]
     [java] 18:29:39,010 actions.LogLocalVariables             WARN  -     Global Name: exampleTableVariable2, value: {First siding={Turnouts=IT15, Sensors=IS15}, Left={Turnouts=Left turnout, Sensors=Left sensor}, Second siding={Turnouts=IT22, Sensors=IS22}, Right={Turnouts=Right turnout, Sensors=Right sensor}} [JMRI LogixNGThread]
     [java] 18:29:39,010 actions.LogLocalVariables             WARN  - Log local variables done [JMRI LogixNGThread]
     [java] 18:29:39,010 actions.LogData                       WARN  - ---- Finish :: IQ:AUTO:0002 ::IQC:AUTO:0004 [JMRI LogixNGThread]

```

IQ:AUTO:0002 is running after IQ:AUTO:0001 even though IQ:AUTO:0002 is in the initialization table.

[DS Test.txt](https://github.com/JMRI/JMRI/files/8587235/DS.Test.txt)

dsand47 commented:
> It turned out that I had a scrollpane in a JPanel in the contentpane. By removing the JPanel, the scrollpane now expands as the dialog is resized. So it was easy to fix.

Thanks.  It looks like the "Log data" action needs the same fix.  Also "Show dialog" and probably others.

danielb987 commented:
A LogixNG variable can now be initialized to a copy of a LogixNG Table. Select the type `LogixNG Table` and enter the name of the table in the column `Initial value`.

danielb987 commented:
> The failure is caused by a timing/sequencing issue.

Thanks. I still do not understand the problem. Can you test PR #10923 and give me the complete log? That PR prints out more info to the log.

danielb987 commented:
> > It turned out that I had a scrollpane in a JPanel in the contentpane. By removing the JPanel, the scrollpane now expands as the dialog is resized. So it was easy to fix.
> 
> Thanks. It looks like the "Log data" action needs the same fix. Also "Show dialog" and probably others.

I have fixed the actions:
* Listen on beans
* Call module
* Log data
* Show dialog

And the expressions:
* Call module
* Log data

And I have fixed the module edit. The dialog where you enter the parameters a module accepts.

danielb987 commented:
When a LogixNG is created, it isn't executed at once. The LogixNG needs to be disabled and then enabled to execute. This is probably related to this PR and I will fix it later today.

danielb987 commented:
> When a LogixNG is created, it isn't executed at once. The LogixNG needs to be disabled and then enabled to execute. This is probably related to this PR and I will fix it later today.

My fault. The LogixNG behaves just as before. I did a thinking error when I wrote the earlier comment.

This PR is ready for review.

dsand47 commented on java/src/jmri/jmrit/logixng/implementation/DefaultLogixNGManager.java:
```diff
@@ -246,6 +246,21 @@ public void activateAllLogixNGs(boolean runDelayed, boolean runOnSeparateThread)
         // Therefore we create a new thread for this task.
         Runnable runnable = () -> {
+
+            // Activate and execute the initialization LogixNGs first.
```
Should say global variable initialization.  Change not required.

danielb987 replied:
Thanks. Fixed.
